dimod/io/coo.py

```diff
@@ -27,58 +27,70 @@
 where 0, 1 are the variable lables and 2.0 is the bias.
 """
 
+_VARTYPE_HEADER_REGEX = r'^[ \t\f]*#.*?vartype[:=][ \t]*([-_.a-zA-Z0-9]+)'
+"""
+The header should be in the first line and look like
+
+# vartype=SPIN
+
+"""
 
-def dumps(bqm):
+
+def dumps(bqm, vartype_header=False):
     """Dump a binary quadratic model to a string in COOrdinate format."""
-    return '\n'.join(_iter_triplets(bqm))
+    return '\n'.join(_iter_triplets(bqm, vartype_header))
 
 
-def dump(bqm, fp):
+def dump(bqm, fp, vartype_header=False):
     """Dump a binary quadratic model to a string in COOrdinate format."""
-    for triplet in _iter_triplets(bqm):
+    for triplet in _iter_triplets(bqm, vartype_header):
         fp.write('%s\n' % triplet)
 
 
-def loads(s, bqm):
+def loads(s, cls=BinaryQuadraticModel, vartype=None):
     """Load a COOrdinate formatted binary quadratic model from a string."""
-    pattern = re.compile(_LINE_REGEX)
-
-    for line in s.split('\n'):
-        match = pattern.search(line)
+    return load(s.split('\n'), cls=cls, vartype=vartype)
 
-        if match is not None:
-            u, v, bias = int(match.group(1)), int(match.group(2)), float(match.group(3))
-            if u == v:
-                bqm.add_variable(u, bias)
-            else:
-                bqm.add_interaction(u, v, bias)
 
-    return bqm
-
-
-def load(fp, bqm):
+def load(fp, cls=BinaryQuadraticModel, vartype=None):
     """Load a COOrdinate formatted binary quadratic model from a file."""
     pattern = re.compile(_LINE_REGEX)
+    vartype_pattern = re.compile(_VARTYPE_HEADER_REGEX)
 
+    triplets = []
     for line in fp:
-        match = pattern.search(line)
+        triplets.extend(pattern.findall(line))
+
+        vt = vartype_pattern.findall(line)
+        if vt:
+            if vartype is None:
+                vartype = vt[0]
+            elif vartype is not vt[0]:
+                raise ValueError("vartypes from headers and/or inputs do not match")
```

Comment:
I would maybe allow kwarg to override the vartype from file.
Just like you're allowing vartype at the end of file to override vartype at the beginning.

Reply:
In this case I am allowing multiple vartype references but they all must be consistent.

Reply:
What's the point of override if it has to be consistent? :)